File: LeetCode/Medium/0743-network-delay-time/0743-network-delay-time.java
import java.util.Arrays;

class Solution {

    static final int MAX = (int) 1e9;

    static int[][] dists;

    public int networkDelayTime(int[][] times, int n, int k) {
        dists = new int[n + 1][n + 1];

        for (int u = 1; u <= n; u++) {
            Arrays.fill(dists[u], MAX);
            
            dists[u][0] = 0;
            dists[u][u] = 0;
        }

        for (int[] time : times) {
            dists[time[0]][time[1]] = time[2];
        }

        floydWarshall(n);

        return getDelayTime(k);
    }

    void floydWarshall(int n) {
        for (int k = 1; k <= n; k++) {
            for (int u = 1; u <= n; u++) {
                for (int v = 1; v <= n; v++) {
                    dists[u][v] = Math.min(dists[u][v], dists[u][k] + dists[k][v]);
                }
            }
        }
    }

    int getDelayTime(int k) {
        int max = -1;

        for (int w : dists[k]) {
            if (w == MAX) {
                return -1;
            }

            max = Math.max(max, w);
        }

        return max;
    }

}

File: LeetCode/Medium/0743-network-delay-time/0743-network-delay-time.kt
import java.util.*

const val MAX = 1e9.toInt()

class Solution {

    lateinit var edges: Array<ArrayList<Edge>>

    fun networkDelayTime(times: Array<IntArray>, n: Int, k: Int): Int {
        edges = Array(n) { arrayListOf() }

        for ((u, v, w) in times) {
            edges[u - 1] += Edge(v - 1, w)
        }

        return dijkstra(k - 1)
    }

    fun dijkstra(start: Int): Int {
        val dists: Array<Int> = Array(edges.size) { MAX }
        val pq = PriorityQueue<Edge>(compareBy(Edge::w))

        pq += Edge(start, 0)
        dists[start] = 0

        while (pq.isNotEmpty()) {
            val (u, cost) = pq.poll()

            if (cost > dists[u]) {
                continue
            }

            for ((v, w) in edges[u]) {
                if (dists[v] <= cost + w) {
                    continue
                }

                pq += Edge(v, cost + w)
                dists[v] = cost + w
            }
        }

        if (MAX in dists) {
            return -1
        }

        return dists.max()
    }

    data class Edge(val node: Int, val w: Int)
}
